compiler/rustc_ast_passes/src/feature_gate.rs

```diff
@@ -144,15 +144,11 @@ impl<'a> PostExpansionVisitor<'a> {
                 _ => Some(param.ident.span),
             })
             .collect();
-        // FIXME: gate_feature_post doesn't really handle multispans...
-        if !non_lt_param_spans.is_empty() && !self.features.non_lifetime_binders {
-            feature_err(
-                &self.sess.parse_sess,
-                sym::non_lifetime_binders,
-                non_lt_param_spans,
-                crate::fluent_generated::ast_passes_forbidden_non_lifetime_param,
-            )
-            .emit();
+        if !non_lt_param_spans.is_empty() {
```

Comment:
What this will do is emit a feature error for every single span. What I want is for it to be a single feature error with multiple primary spans.

I prefer if this is not changed, or it is fixed in the right way.

Reply:
IDK why there isn't a test for this. Here's one:

```
fn test() -> for<T, S> fn(T, S) { todo!() }
```

```
error[E0658]: only lifetime parameters can be used in this context
 --> src/lib.rs:1:18
  |
1 | fn test() -> for<T, S> fn(T, S) { todo!() }
  |                  ^  ^
  |
  = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
```

Reply:
:+1: 

I have updated the code, removing the invalid commit, and adding three new commits: another small refactor, an alternative fix for the FIXME, and a test update. Hopefully that's closer to what is necessary :)  Those macros still annoy me, but I can't see a better way to do things without increasing the amount of boilerplate.